### ISSUE TITLE
`donut.system/stop-failed-system` throws exception if dependency instance is missing

### ISSUE THREAD

**Issue by steffan-westcott** [donut-party/system#29]
If an exception is thrown during `donut.system/start`, that exception contains the partially started system and may be passed to `donut.system/stop-failed-system` to stop it. However, this stopping procedure does not work if the missing component(s) are dependencies for other components, despite the fact those other components are never started.

I'm using version 0.0.239 of `donut-party/system`. I have discovered two variants of apparently same issue as described above. The following example namespace shows both of them:

```clojure
(ns example.donut-issue
  (:require [donut.system :as ds]))

(def two-component-system
  {::ds/defs {:group {:component-b {::ds/start (fn [_] (throw (RuntimeException. "Error starting component-b")))}
                      :component-c {::ds/config {:component-b (ds/local-ref [:component-b])}
                                    ::ds/start (fn [_] (println "Starting component-c"))}}}})

(def three-component-system
  {::ds/defs {:group {:component-a {::ds/start (fn [_] (println "Starting component-a"))}
                      :component-b {::ds/config {:component-a (ds/local-ref [:component-a])}
                                    ::ds/start  (fn [_] (throw (RuntimeException. "Error starting component-b")))}
                      :component-c {::ds/config {:component-b (ds/local-ref [:component-b])}
                                    ::ds/start  (fn [_] (println "Starting component-c"))}}}})

(defn start [system]
  (try
    (ds/start system)                 ; Should fail to start
    (catch Exception e
      (print (ex-data e))             ; Correctly complains start signal could not be applied to [:group :component-b] ...
      (ds/stop-failed-system e)       ; ...but throws another exception instead of stopping the system (see below)
      )))

(comment
  ;; Fails with this exception:
  ;; Execution error (ExceptionInfo) at donut.system/group-ref-exception (system.cljc:304).
  ;; Invalid group ref: '[:group :component-b]' references undefined component group ':group'
  (start two-component-system)

  ;; Fails with this exception:
  ;; Execution error (ExceptionInfo) at donut.system/ref-exception (system.cljc:297).
  ;; Invalid ref: '[:group :component-c]' references undefined component '[:group :component-b]
  (start three-component-system)
  ;
  )
```

I've found a consequence of this issue is that `donut.system.repl/start` often does not work as intended if the system fails to start, as this uses `donut.system/stop-failed-system`. The exception from `donut.system.repl/start` is the one thrown during stopping rather than starting the system, which was unexpected and frustrating as I had to look elsewhere to debug why the system was not starting correctly.

**Comment by flyingmachine** [comment]
Sorry you ran into this! Thank you for reporting this

**Comment by steffan-westcott** [comment]
Fixed as of version 0.0.241, thank you 😊
